src/data/list/chain_dup.lean

```diff
@@ -0,0 +1,126 @@
+/-
+Copyright (c) 2022 Eric Rodriguez. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Rodriguez, Eric Wieser
+-/
+import data.list.chain
+
+/-!
+# Chain Deduplication of Lists
+
+This file defines `list.chain'_dedup`, which removes all duplicates that are adjacent in a list,
+e.g. `[2, 2, 3, 3, 2].chain'_dedup = [2, 3, 2]`.
+
+## Main definitions
+
+* `list.chain'_dedup`: removes all adjacent duplicates in a list.
+
+## Main statements
+
+* `list.chain'_dedup_sublist`: `l.chain'_dedup` is a sublist of `l`.
+* `list.chain'_dedup_is_chain'`: `l.chain'_dedup` satisfies `chain' ne`.
+
+## Tags
+
+adjacent, chain, duplicates, remove, list
+-/
+
+
+variables {α : Type*} [decidable_eq α] (a b : α) (l : list α)
+
+namespace list
+
+/-- Removes all adjacent duplicates in `a :: l`. -/
+def chain_dedup : α → list α → list α
+| a [] := []
+| a (h :: l) := if a = h then chain_dedup h l else h :: chain_dedup h l
```

Comment:
This and `chain_dedup'` should be in `data.list.defs`

Reply:
and on 3, it seems to break all my proofs in ways that make it seem genuinely different from what I'd expect - if I'm honest, I focused mainly on `chain'_dedup`, so I didn't pay as close attention as to how exactly to implement `chain_dedup` without it. Do you think the base case should be `| a [] := [a]`? That seems reasonable, now that I think about it a little

Reply:
Note that with that generalization, this is the `chain` version of [`list.pw_filter`](https://leanprover-community.github.io/mathlib_docs/data/list/defs.html#list.pw_filter)